pennylane/plugins/expt_tensornet.py

```diff
@@ -0,0 +1,197 @@
+        'BasisState': None,
+        'QubitUnitary': unitary,
+        'PauliX': X,
+        'PauliY': Y,
+        'PauliZ': Z,
+        'Hadamard': H,
+        'S': S,
+        'T': T,
+        'CNOT': CNOT,
+        'SWAP': SWAP,
+        'CSWAP':CSWAP,
+        'CZ': CZ,
+        'PhaseShift': Rphi,
+        'RX': Rotx,
+        'RY': Roty,
+        'RZ': Rotz,
+        'Rot': Rot3,
+        'CRX': CRotx,
+        'CRY': CRoty,
+        'CRZ': CRotz,
+        'CRot': CRot3
+    }
+
+    _observable_map = {
+        'PauliX': X,
+        'PauliY': Y,
+        'PauliZ': Z,
+        'Hadamard': H,
+        'Hermitian': hermitian,
+        'Identity': identity
+    }
+
+    def __init__(self, wires):
+        super().__init__(wires, shots=1)
+        self.eng = None
+        self.analytic = True
+        self._nodes = []
+        self._edges = []
+        self._zero_state = np.zeros([2] * wires)
+        self._zero_state[tuple([0] * wires)] = 1.0
+        # TODO: since this state is separable, can be more intelligent about not making a dense matrix
+        self._state = self._add_node(self._zero_state, wires=tuple(w for w in range(wires)), name="AllZeroState")
+        self._free_edges = [e for e in self._state.edges]  # we need this list to be distinct from self._state.edges
+
+    def _add_node(self, A, wires, name="UnnamedNode"):
+        name = "{}{}".format(name, tuple(w for w in wires))
+        if isinstance(A, tn.Node):
+            A.set_name(name)
+            node = A
+        else:
+            node = tn.Node(A, name=name)
+        self._nodes.append(node)
+        return node
+
+    def _add_edge(self, node1, idx1, node2, idx2):
+        edge = tn.connect(node1[idx1], node2[idx2])
+        self._edges.append(edge)
+        return edge
+
+    def pre_apply(self):
+        self.reset()
+
+    def apply(self, operation, wires, par):
+        A = self._get_operator_matrix(operation, par)
+        num_mult_idxs = len(wires)
+        A = np.reshape(A, [2] * num_mult_idxs * 2)
+        op_node = self._add_node(A, wires=wires, name=operation)
+        for idx, w in enumerate(wires):
+            self._add_edge(op_node, num_mult_idxs + idx, self._state, w)
+            self._free_edges[w] = op_node[idx]
+        # TODO: can be smarter here about collecting contractions?
+        self._state = tn.contract_between(op_node, self._state, output_edge_order=self._free_edges)
+        print(self._state._tensor.reshape(-1))
+        print("\n")
+
+    def expval(self, observable, wires, par):
+        if not isinstance(observable, list):
+            observable = [observable]
+            wires = [wires]
+            par = [par]
+        matrices = []
+        for o, p, w in zip(observable, par, wires):
+            A = self._get_operator_matrix(o, p)
+            num_mult_idxs = len(w)
+            matrices.append(np.reshape(A, [2] * num_mult_idxs * 2))
+        nodes = [self._add_node(A, w, name=o) for A, w, o in zip(matrices, wires, observable)]
```

Comment:
So just to clarify here, we are creating a list of matrices we wish to measure, adding them to the network, and then sending them to `self.ev` for contraction?

Reply:
Yes, except the nodes aren't connected to the network until inside `self.ev`